### PR TITLE
[VUFIND-1614] Modernize PubDateVisAjax recommendation module

### PR DESCRIPTION
**Pull request by ThoWagen**
In this PR a combination of charts.js and the date range slider from  #4238 is used to replace PubDateVisAjax solution using the jQuery Flot library. The aim was also to increase accessibility and general user experience by using a slider and input fields instead of manipulating the chart with the mouse.

- [x] Resolve [VUFIND-1614](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1614) when merging
- [x] Update changelog when merging

### CODE REVIEW

**Review by sturkel89 — CHANGES_REQUESTED** [review]
I think this interface is a nice improvement over the existing tool. When it's activated, the user can be very specific about the date range they want to see via the sliders or the text entry boxes. 

In the existing tool, the user can only drag to select, it's not clear how to do that, and the selection is very imprecise. Also, if the user wants to change their selection, they can't adjust on one end only; they need to redraw the entire selection.

In any case, the new tool improves functionality! But I have a couple of critiques and questions. 

NB: the universe of publication dates in the test data goes from **1783 through 2016**.

**Test:**
image

**Dev:**
image

--------------

### True/False setting is not functional

In searches.ini, this is how Demian suggested I set up this tool:

`default_top_recommend[] = "PubDateVisAjax:true:publishDate" 
`

With the option set to 'true,' when the user selects a date range the timeline zooms in and only shows the specified dates plus a buffer of five years outside those boundaries:

****Test****
image

**Dev**
image

If the option is changed to 'false' on test, the zooming in doesn't happen. The user selects a date range, and the graphic shows the shaded area for the date range of interest in the context of the full date range for the entire universe of materials. 

This 'false' setting is functional in the dev branch, but not in the new tool on the test branch.

**Test (display is the same with setting at true or false)**
image

**Dev**
image

--------------------

### Visibility of text entry boxes and Set and Clear buttons

I don't like that the controls (text entry boxes and set and clear buttons) are only visible after you begin interacting with the slider, and also disappear after you clear your selections. 

I understand that it saves vertical space to hide those controls when the user is not interested in interacting with the slider, so I'm not sure what the best solution is here. Could there be an option in searches.ini that would allow the library to specify whether the controls are always visible vs. only visible during use?

Or maybe the default should be that the actual years covering the collection are selected with the shading and data entry boxes, and all of the parts of the tool are visible at all times? The current default seems colorless when you compare it with the tool when it's "in action."

---------------

### Year-specific publications

It's a nice touch that the user can mouse over a specific bar in the graph to see how many items were published in that particular year. However, it's frustrating that the bar and/or pop-up are not clickable to limit the results to ONLY the nine items published in 1992.

image

----------------

### General aesthetics

1. I prefer the shorter height of the date range display on dev. 

2. I would prefer a more attractive color for the shading, or the option to specify the color. The brown is unappealing to my eye.

3. I like the faint vertical lines that appear at 50 year intervals on the dev version. The blank field behind the timeline on test is a little too vast. 

4. Would it be possible to emulate dev in displaying the year range in "round" numbers? It would be nice to always have these more standardized years displayed on the graphic. You can see the examples in the screenshots above; here's a textual description of the differences between test and dev:

   - The labels on dev before limits are applied show 1800, 1850, 1900, 1950, and 2000. The labels on test show 1778, 1803, 1828, 1853, .... 1953, 1978, 2003. 

   - After limits are applied, dev still displays the graph with labels at every five years (1970 1975 1980 ... 2010 2015 2020) while test now uses a different set of non-round numbers (1970 1976 1982 1988 1994 2000 2006 2012 2018)

**Comment by ThoWagen** [comment]
Thanks @sturkel89, those are some good points!

**True/False setting is not functional**
The zoom setting is now fixed.

**Visibility of text entry boxes and Set and Clear buttons**
I also added the option to always display the controls as you suggested. You can control this by setting "true" or "false" after the zoom setting, i.e.

Always show controls
`default_top_recommend[]  = PubDateVisAjax:true:false:publish_year`

Hide controls initially
`default_top_recommend[]  = PubDateVisAjax:true:true:publish_year`

**Year-specific publications**
If you click in the chart, the selection will be reduced to the respective year.

**General aesthetics**
1. The height of the chart is now reduced. This can also be customized in own themes.
2. I changed the default color to blue. Do you like this more? This is also easily customizable in custom themes.
3. The grid is now being displayed.
4. We need to write some own logic for that but now the labels should be nicely distributed.

**Review by sturkel89 — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen ! I think this is looking better and better. It is almost perfect!

I like the new more vertically compact layout, the "round" years, and the presence of tick marks. I also see that zooming works well.

Here are a couple of screenshots:

**Bootstrap5:**

image

**Sandal5:**

image

---------------

### Tick marks

My critique of the tick marks is that I'd prefer if they lined up with the years instead of being between years. I am inclined to create my custom date range by dragging from tick mark to tick mark, but then I end up with "odd" years selected. This is the selection I made that resulted in the outcome above:

image

Could the tick marks be aligned with the year labels, as they are on dev?

-----------------------------

### Colors

I like the blue much better than the brown! I think the color looks great in bootstrap5. 

I think the color seems a little bit intense in sandal5 since there are other strong colors there in the header -- however, it's not terrible. It might look better if it repeated one of the lighter colors that's already in use on the page -- from the breadcrumbs bar, or background of the Reset Filters button, or the background of the sidebar.

---------------------------

### Settings confusion

You showed this example of the settings line with controls always displayed:

`default_top_recommend[]  = PubDateVisAjax:true:false:publish_year`

I have been using this setting, based on Demian's suggestion to me when I first checked the PR:

`default_top_recommend[] = PubDateVisAjax:true:publishDate`

I'm finding that if I change the last bit to publish_year, I don't see the slider at all. Only publishDate works for me. Is this expected?

(I see that the true:true setting hides the text entry and set/clear buttons initially, and the true:false setting shows them all the time, so that improvement works!)

**Comment by ThoWagen** [comment]
The grid is now aligned with the labels and I adjusted the color in sandal5. What do you think @sturkel89?

And we use a different facet name locally so I accidentally used that in the example. With the standard VuFind index you should use `default_top_recommend[]  = PubDateVisAjax:true:false:publishDate` and `default_top_recommend[]  = PubDateVisAjax:true:false:publishDate` instead as you probably have already done. Sorry for the confusion.

**Review by sturkel89 — COMMENTED** [review]
Thanks for making these adjustments, @ThoWagen! The tick marks now align with the "round" years, and it's easy to use the sliders to select those round years. The gray color in Sandal5 looks better to me than the more intense blue.

image


You've done so much, so I hesitate to ask for any other changes. I'll report my TINY new thought/question: could the bars in the graph be solid instead of outlined? I think the feature looks a bit busy the way it is, and would be easier on the eyes if the bars were filled in with the same solid color as the outline.

I'll submit this as a comment and will be happy to change to 'approve' after hearing back from you. Thanks!

**Comment by ThoWagen** [comment]
@sturkel89 I agree that solid bars looks better. I only changed it for sandal5 for now, since I was not sure if you only meant that. 

And please feel always free to always suggest improvements! I'm happy to invest the work if there is room for improvement :)

**Review by sturkel89 — CHANGES_REQUESTED** [review]
That looks great!!

image

I would suggest making the same change to bootstrap5 as well. It would surely be an improvement.

image

I'm not sure what you were planning to do, but in my opinion the default setting should be:

`default_top_recommend[] = PubDateVisAjax:true:false:publishDate`

I think this feature will be completely ready to merge once the bars are filled in in the second theme. Thanks so much!

**Comment by ThoWagen** [comment]
@sturkel89 Now the bars are also filled in the bootstrap5 theme.

The default for zooming has already been "false" and I don't think we should change that now. But the controls are always visible by default.

**Review by sturkel89 — APPROVED** [review]
Looks great!

image

I understand your preference to not change the default. I hope the libraries that adopt this tool will consider the options and make the best decision. In any case, as far as I'm concerned, this is ready to merge!

File: themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml [hunk 1]
```diff
@@ -1,27 +1,54 @@
...
   ?>
   <?php foreach ($visFacets as $facetField => $facetRange): ?>
     <div class="authorbox">
-      <div id="datevis<?=$this->escapeHtml($facetField)?>xWrapper" class="hidden">
-        <strong><?=$this->transEsc($facetRange['label']) ?></strong>
-        <?php /* space the flot visualisation */ ?>
-        <div id="datevis<?=$facetField ?>x" style="margin:0 10px;width:auto;height:80px;cursor:crosshair;"></div>
-        <div id="clearButtonText" style="display: none"><?=$this->transEsc('Clear') ?></div>
-      </div>
+      <form class="datevis-form facet-range-form hidden" name="datevis<?=$facetField?>xForm" id="datevis<?=$facetField?>xForm" data-name="<?=$this->escapeHtmlAttr($facetField)?>" data-filter-field="daterange[]">
```
Comment (demiankatz) [inline themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: Is there a reason that `$facetField` is only escaped in `data-name`, but not in the `name` or `id` attributes (or in several subsequent references below)? Maybe we should make a `$facetFieldEsc` variable to straighten this out, or switch to using the `htmlAttributes` view helper for consistency.

File: config/vufind/searches.ini [hunk 2]
```diff
@@ -493,11 +493,13 @@ pattern[sort_first_indexed_desc] = "^first_indexed desc$"
...
-; PubDateVisAjax:[zooming]:[facet field 1]:[facet field 2]:...:[facet field n]
+; PubDateVisAjax:[zooming]:[initiallyHideControls]:[facet field 1]:[facet field 2]:...:[facet field n]
 ;       Display a visualization of publication dates for each of the specified facet
-;       fields.  This is designed for a field containing four-digit years.  Zooming
+;       fields.  This is designed for a field containing four-digit years. Zooming
 ;       is set to either true or false and controls whether or not selecting part of
-;       the graph zooms in to greater detail.
+;       the graph zooms in to greater detail. InitiallyHideControls is set to either
+;       true or false and controls whether or not the complete controls should be visible
+;       even if the slider was not interacted with.
```
Comment (demiankatz) [inline config/vufind/searches.ini]: While we're updating the comments here, we should probably indicate the default values of the two boolean arguments.

File: module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php [hunk 3]
```diff
@@ -44,33 +44,42 @@
...
      * @var \VuFind\Search\Base\Results
      */
-    protected $searchObject;
+    protected \VuFind\Search\Base\Results $searchObject;
 
     /**
      * Should we allow zooming? (String of "true" or "false")
      *
-     * @var string
+     * @var bool
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php:66]: The comment here still says "string" -- the parenthetical comment can probably be removed.

File: module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php [hunk 4]
```diff
@@ -44,33 +44,42 @@
...
      */
-    protected $zooming;
+    protected bool $zooming = false;
+
+    /**
+     * Should we initially hide controls? (String of "true" or "false")
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php]: Similarly, the string comment doesn't really seem accurate.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen, for all the improvements, and @sturkel89 for the thorough testing and feedback!

See below for a few minor suggestions from a review of the code.

File: themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml (hunk 1 above)
Comment (ThoWagen) [reply themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: No, there is not reason. I created `$facetFieldEsc` because it is also used in all of the other elements.

File: config/vufind/searches.ini (hunk 2 above)
Comment (ThoWagen) [reply config/vufind/searches.ini]: Done

File: module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php (hunk 3 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php:66]: Ah yes, of course. Fixed

File: module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php (hunk 4 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php]: Also fixed.

File: themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml [hunk 5]
```diff
@@ -23,26 +23,29 @@
...
   <?php foreach ($visFacets as $facetField => $facetRange): ?>
+    <?php
+      $facetFieldEsc = $this->escapeHtmlAttr($facetField);
+    ?>
     <div class="authorbox">
-      <form class="datevis-form facet-range-form hidden" name="datevis<?=$facetField?>xForm" id="datevis<?=$facetField?>xForm" data-name="<?=$this->escapeHtmlAttr($facetField)?>" data-filter-field="daterange[]">
-        <?=$this->results->getUrlQuery()->asHiddenFields(['page' => '/./', 'filter' => "/^{$facetField}:.*/"])?>
-        <input type="hidden" name="daterange[]" value="<?=$this->escapeHtmlAttr($facetField)?>">
+      <form class="datevis-form facet-range-form hidden" name="datevis<?=$facetFieldEsc?>xForm" id="datevis<?=$facetFieldEsc?>xForm" data-name="<?=$facetFieldEsc?>" data-filter-field="daterange[]">
+        <?=$this->results->getUrlQuery()->asHiddenFields(['page' => '/./', 'filter' => "/^{$facetFieldEsc}:.*/"])?>
```
Comment (demiankatz) [inline themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: I don't think we want to use $facetFieldEsc here, since this is a regex rather than HTML.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen -- nearly done, but I think there might be just one case of using the wrong kind of escaping in the wrong context.

File: themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml (hunk 5 above)
Comment (ThoWagen) [reply themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: Sorry, don't know how I overlooked that. Fixed.

Comment (ThoWagen) [reply themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: Actually, should we use `preg_quote` since that is a regex?

Comment (demiankatz) [reply themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: Yes, I think `preg_quote` would be appropriate in this context. Thanks!

Comment (ThoWagen) [reply themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: Alright, I added it

File: themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml [hunk 6]
```diff
@@ -1,27 +1,57 @@
...
+    ?>
     <div class="authorbox">
-      <div id="datevis<?=$this->escapeHtml($facetField)?>xWrapper" class="hidden">
-        <strong><?=$this->transEsc($facetRange['label']) ?></strong>
-        <?php /* space the flot visualisation */ ?>
-        <div id="datevis<?=$facetField ?>x" style="margin:0 10px;width:auto;height:80px;cursor:crosshair;"></div>
-        <div id="clearButtonText" style="display: none"><?=$this->transEsc('Clear') ?></div>
-      </div>
+      <form class="datevis-form facet-range-form hidden" name="datevis<?=$facetFieldEsc?>xForm" id="datevis<?=$facetFieldEsc?>xForm" data-name="<?=$facetFieldEsc?>" data-filter-field="daterange[]">
+        <?=$this->results->getUrlQuery()->asHiddenFields(['page' => '/./', 'filter' => '/^{' . preg_quote($facetField) . '}:.*/'])?>
```
Comment (demiankatz) [inline themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: It looks like the braces used for embedding a variable in a double-quoted string got converted into literal braces in single-quoted strings, which broke the regular expression.

```suggestion
        <?=$this->results->getUrlQuery()->asHiddenFields(['page' => '/./', 'filter' => '/^' . preg_quote($facetField) . ':.*/'])?>
```

**Review by demiankatz — CHANGES_REQUESTED** [review]
@ThoWagen, I'm seeing one problem: every time I adjust the timeline selection and click "Set" it adds another separate date range filter. I believe the expected behavior is that the most recent range should replace any/all prior ranges. Otherwise, things can get very confusing!

I'm pretty sure the fix below will address this. In fact, I'm going to commit the change after submitting this review, and if it sorts everything out, I think I can go ahead and approve the work. :-)

**Review by demiankatz — APPROVED** [review]
My fix seems to have worked! Thanks again for working on this, @ThoWagen!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Actually, I found one more last-minute issue. It occurred to me that while this component is generally designed for use in the top area, it is sometimes valuable to include it in the sidebar instead -- [at Villanova, we include it embedded in the facet list](https://library.villanova.edu/Find/Search/Results), so the graph is next to the date facet.

Unfortunately, if I try that configuration with the new control, the inputs get so squashed that you can't see their contents:

image

Can we do something about this -- either wrap the buttons to another line or otherwise make room for usable text inputs?

**Comment by demiankatz** [comment]
(On a related note, I also notice that the entire graph disappears if you include this in the top area on a mobile device, leaving only the slider control visible; maybe that's a good thing, but I'm not entirely sure...).

File: themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml (hunk 6 above)
Comment (ThoWagen) [reply themes/bootstrap5/templates/Recommend/PubDateVisAjax.phtml]: Sorry, for missing that again and thank you for fixing it!

**Comment by ThoWagen** [comment]
Thanks @demiankatz, I did not think about the sidebar. That is adjusted now. 

The hiding of the chart on mobile devices was intentional. But on second thought I believe that only having the input also does not make sense. The best way to avoid that the recommendation module does take to much space on mobile would probably be to hide it completely and instead add it or a simple facet to the sidebar.

So I removed the hiding of the chart on mobile and we can leave it as is for everyone to customize it as preferred.

**Comment by ThoWagen** [comment]
I just noticed a problem when having multiple version with different zooming or 'initially-hide-controls' settings. Only one of them gets applied. I will fix that.

**Comment by ThoWagen** [comment]
The problem is fixed now.

**Review by demiankatz — APPROVED** [review]
Thanks for the fixes (and the last-minute catch). All looks good to me now!
